torchani/cuaev/aev.cu

```diff
@@ -574,11 +679,167 @@ torch::Tensor cuComputeAEV(
+                              res.tensor_angularRij,
+                              EtaR_t,
+                              ShfR_t,
+                              EtaA_t,
+                              Zeta_t,
+                              ShfA_t,
+                              ShfZ_t});
+      ctx->saved_data["aev_params"] = res.aev_params.toTuple();
+      c10::List<int64_t> int_list({res.total_natom_pairs, res.nRadialRij, res.nAngularRij});
+      ctx->saved_data["int_list"] = torch::IValue(int_list);
+    }
+    return res.aev_t;
+  }
+
+  static tensor_list backward(torch::autograd::AutogradContext* ctx, tensor_list grad_outputs) {
+    printf("calling autograd::backward \n");
+    auto saved = ctx->get_saved_variables();
+    auto coordinates_t = saved[0];
+    auto species_t = saved[1];
+    auto tensor_Rij = saved[2];
+    auto tensor_radialRij = saved[3];
+    auto tensor_angularRij = saved[4];
+    auto EtaR_t = saved[5], ShfR_t = saved[6], EtaA_t = saved[7], Zeta_t = saved[8], ShfA_t = saved[9],
+         ShfZ_t = saved[10];
+
+    c10::intrusive_ptr<c10::ivalue::Tuple> aev_params_tp_ptr = ctx->saved_data["aev_params"].toTuple();
+    auto aev_params_tp = aev_params_tp_ptr->elements();
+    AEVScalarParams<float> aev_params = {(float)aev_params_tp[0].toDouble(),
+                                         (float)aev_params_tp[1].toDouble(),
+                                         (int)aev_params_tp[2].toInt(),
+                                         (int)aev_params_tp[3].toInt(),
+                                         (int)aev_params_tp[4].toInt(),
+                                         (int)aev_params_tp[5].toInt(),
+                                         (int)aev_params_tp[6].toInt()};
+    c10::List<int64_t> int_list = ctx->saved_data["int_list"].toIntList();
+    int total_natom_pairs = (int)int_list[0];
+    int nRadialRij = (int)int_list[1];
+    int nAngularRij = (int)int_list[2];
+
+    Tensor grad_coord = cuaev_backward(
+        grad_outputs[0],
+        coordinates_t,
+        species_t,
+        aev_params,
+        EtaR_t,
+        ShfR_t,
+        EtaA_t,
+        Zeta_t,
+        ShfA_t,
+        ShfZ_t,
+        tensor_Rij,
+        total_natom_pairs,
+        tensor_radialRij,
+        nRadialRij,
+        tensor_angularRij,
+        nAngularRij);
+
+    return {
+        grad_coord, Tensor(), Tensor(), Tensor(), Tensor(), Tensor(), Tensor(), Tensor(), Tensor(), Tensor(), Tensor()};
+  }
+};
+
+Tensor cuaev_autograd(AEV_INPUT) {
+  printf("calling autograd \n");
+  return CuaevAutograd::apply(
+      coordinates_t, species_t, Rcr_, Rca_, EtaR_t, ShfR_t, EtaA_t, Zeta_t, ShfA_t, ShfZ_t, num_species_);
 }
 
 TORCH_LIBRARY(cuaev, m) {
-  m.def("cuComputeAEV", &cuComputeAEV<float>);
+  m.def("cuComputeAEV", cuaev_cuda);
+}
+
+TORCH_LIBRARY_IMPL(cuaev, CUDA, m) {
+  m.impl("cuComputeAEV", cuaev_cuda);
+}
+
+TORCH_LIBRARY_IMPL(cuaev, Autograd, m) {
+  m.impl("cuComputeAEV", cuaev_autograd);
```

Comment:
Does this work?
```C++
m.impl("cuComputeAEV", CuaevAutograd::apply);
```

Reply:
The official dispatcher example is using the same style, so I guess it's good? https://github.com/pytorch/tutorials/blob/master/advanced_source/dispatcher/op.cpp#L77